### PR TITLE
Update installer support for main

### PR DESCRIPTION
**Pull request by lstein**
#  Make InvokeAI package installable by mere mortals
    
This commit makes InvokeAI 3.0 to be installable via PyPi.org and/or the installer script. The install process is now pretty much identical to the 2.3 process, including creating launcher scripts `invoke.sh` and `invoke.bat`.
    
Main changes:
    
1. Moved static web pages into `invokeai/frontend/web` and modified the API to look for them there. This allows pip to copy the files into the distribution directory so that user no longer has to be in repo root to launch, and enables PyPi installations with `pip install invokeai`
    
2. Update invoke.sh and invoke.bat to launch the new web application properly. This also changes the wording for launching the CLI from "generate images" to "explore the InvokeAI node system," since I would not recommend using the CLI to generate images routinely.
    
3. Fix a bug in the checkpoint converter script that was identified during testing.
    
4. Better error reporting when checkpoint converter fails.
    
5. Rebuild front end.

# Major improvements to the model installer.

1. The text user interface for `invokeai-model-install` has been expanded to allow the user to install controlnet, LoRA, textual inversion, diffusers and checkpoint models. The user can install interactively (without leaving the TUI), or in batch mode after exiting the application.
 
image

2. The `invokeai-model-install` command now lets you list, add and delete models from the command line:

## Listing models
```
$ invokeai-model-install --list diffusers
Diffuser models:
analog-diffusion-1.0      not loaded  diffusers  An SD-1.5 model trained on diverse analog photographs (2.13 GB)
d&d-diffusion-1.0         not loaded  diffusers  Dungeons & Dragons characters (2.13 GB)
deliberate-1.0            not loaded  diffusers  Versatile model that produces detailed images up to 768px (4.27 GB)
DreamShaper               not loaded  diffusers  Imported diffusers model DreamShaper
sd-inpainting-1.5         not loaded  diffusers  RunwayML SD 1.5 model optimized for inpainting, diffusers version (4.27 GB)
sd-inpainting-2.0         not loaded  diffusers  Stable Diffusion version 2.0 inpainting model (5.21 GB)
stable-diffusion-1.5      not loaded  diffusers  Stable Diffusion version 1.5 diffusers model (4.27 GB)
stable-diffusion-2.1      not loaded  diffusers  Stable Diffusion version 2.1 diffusers model, trained on 768 pixel images (5.21 GB)
```

```
$ invokeai-model-install --list tis
Loading Python libraries...

Installed Textual Inversion Embeddings:
   EasyNegative
   ahx-beta-453407d
```

## Installing models

(this example shows correct handling of a server side error at Civitai)
```
$ invokeai-model-install --diffusers https://civitai.com/api/download/models/46259 Linaqruf/anything-v3.0
Loading Python libraries...

[2023-06-05 22:17:23,556]::[InvokeAI]::INFO --> INSTALLING EXTERNAL MODELS
[2023-06-05 22:17:23,557]::[InvokeAI]::INFO --> Probing https://civitai.com/api/download/models/46259 for import
[2023-06-05 22:17:23,557]::[InvokeAI]::INFO --> https://civitai.com/api/download/models/46259 appears to be a URL
[2023-06-05 22:17:23,763]::[InvokeAI]::ERROR --> An error occurred during downloading /home/lstein/invokeai-test/models/ldm/stable-diffusion-v1/46259: Internal Server Error
[2023-06-05 22:17:23,763]::[InvokeAI]::ERROR --> ERROR DOWNLOADING https://civitai.com/api/download/models/46259: {"error":"Invalid database operation","cause":{"clientVersion":"4.12.0"}}
[2023-06-05 22:17:23,764]::[InvokeAI]::INFO --> Probing Linaqruf/anything-v3.0 for import
[2023-06-05 22:17:23,764]::[InvokeAI]::DEBUG --> Linaqruf/anything-v3.0 appears to be a HuggingFace diffusers repo_id
[2023-06-05 22:17:23,768]::[InvokeAI]::INFO --> Loading diffusers model from Linaqruf/anything-v3.0
[2023-06-05 22:17:23,769]::[InvokeAI]::DEBUG --> Using faster float16 precision
[2023-06-05 22:17:23,883]::[InvokeAI]::ERROR --> An unexpected error occurred while downloading the model: 404 Client Error. (Request ID: Root=1-647e9733-1b0ee3af67d6ac3456b1ebfc)

Revision Not Found for url: https://huggingface.co/Linaqruf/anything-v3.0/resolve/fp16/model_index.json.
Invalid rev id: fp16)
Downloading (…)ain/model_index.json: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 511/511 [00:00<00:00, 2.57MB/s]
Downloading (…)cial_tokens_map.json: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 472/472 [00:00<00:00, 6.13MB/s]
Downloading (…)cheduler_config.json: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 341/341 [00:00<00:00, 3.30MB/s]
Downloading (…)okenizer_config.json: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 807/807 [00:00<00:00, 11.3MB/s]
```

## Deleting models

```
 invokeai-model-install --delete --diffusers anything-v3
Loading Python libraries...

[2023-06-05 22:19:45,927]::[InvokeAI]::INFO --> Processing requested deletions
[2023-06-05 22:19:45,927]::[InvokeAI]::INFO --> anything-v3...
[2023-06-05 22:19:45,927]::[InvokeAI]::INFO --> Deleting the cached model directory for Linaqruf/anything-v3.0
[2023-06-05 22:19:45,948]::[InvokeAI]::WARNING --> Deletion of this model is expected to free 4.3G

```

### CODE REVIEW

**Comment by psychedelicious** [comment]
Does this mean `pip install InvokeAI` will get the v3 package?

**Comment by lstein** [comment]
> Does this mean `pip install InvokeAI` will get the v3 package?

Only when we make a release on GitHub.

File: invokeai/app/api_app.py [hunk 1]
```diff
@@ -120,7 +122,7 @@ def custom_openapi():
 app.openapi = custom_openapi
 
 # Override API doc favicons
-app.mount("/static", StaticFiles(directory="static/dream_web"), name="static")
+app.mount("/static", StaticFiles(directory=Path(web_dir.__path__[0], 'static/dream_web')), name="static")
```
Comment (psychedelicious) [inline invokeai/app/api_app.py:125]: Where does `web_dir` come from? It doesn't exist as far as I can tell

File: invokeai/app/api_app.py [hunk 2]
```diff
@@ -142,10 +144,11 @@ def overridden_redoc():
 
 
 # Must mount *after* the other routes else it borks em
-app.mount(
-    "/", StaticFiles(directory="invokeai/frontend/web/dist", html=True), name="ui"
-)
-
+app.mount("/", 
+          StaticFiles(directory=Path(web_dir.__path__[0],"dist"),
```
Comment (psychedelicious) [inline invokeai/app/api_app.py:148]: Same comment re: `web_dir` not existing

File: invokeai/backend/config/model_install_backend.py [hunk 3]
```diff
@@ -149,10 +149,8 @@ def yes_or_no(prompt: str, default_yes=True):
 def get_root(root: str = None) -> str:
     if root:
         return root
-    elif os.environ.get("INVOKEAI_ROOT"):
-        return os.environ.get("INVOKEAI_ROOT")
     else:
-        return config.root
+        return config.root_dir
```
Comment (psychedelicious) [inline invokeai/backend/config/model_install_backend.py:153]: `config.root_dir` appears to be a `Path`, but the function typing indicates it always returns a string.

**Review by psychedelicious — CHANGES_REQUESTED** [review]
Couple changes requested

**Review by psychedelicious — CHANGES_REQUESTED** [review]

File: invokeai/backend/config/model_install_backend.py (hunk 3 above)
Comment (psychedelicious) [inline invokeai/backend/config/model_install_backend.py:153]: Think my comment here didn't survive - `config.root_dir` is a `Path` object, but this function indicates its return type is always `str`. Let's update the return type, or cast the `Path` to string when returning, whichever is appropriate.

File: invokeai/app/api_app.py (hunk 1 above)
Comment (lstein) [reply invokeai/app/api_app.py:127]: So sorry. This should never have worked and I don't understand how it seemed to work when I tested it. I have added the missing line:

```import invokeai.frontend.web as web_dir```

File: invokeai/app/api_app.py (hunk 2 above)
Comment (lstein) [reply invokeai/app/api_app.py:149]: See above. There was a missing import statement.

File: invokeai/backend/config/model_install_backend.py (hunk 3 above)
Comment (lstein) [reply invokeai/backend/config/model_install_backend.py:153]: Even better. The whole `get_root()` function is never used, so I got rid of it.

**Review by psychedelicious — APPROVED** [review]
Looks good. Two notes:
- I needed to run `yarn vite build`, it looks like the javascript was there but the `index.html` file wasn't correct.
- This now points the runtime dir back to what is specified in `invokeai.init`. For me, because I had at some point changed the schema of that database, I had to delete the database. Alternatively, I could have dropped the invalid `images` table.

But everything does work!

**Comment by lstein** [comment]
H'mmm. `invokeai.init` shouldn't be read any longer. It was replaced by
`invokeai.yaml` when the new configuration system was merged.

I'm also aware that there's still some confusion regarding where the
`outputs` directory (which contains `invokeai.db`) is being placed. In
current main, `outputs` and `outputs/invokeai.db` is created relative to
the repo root. In PR #3448 `outputs` gets moved into
`INVOKEAI_ROOT/outputs`. I'm actually not sure that it makes sense for the
database to be in `outputs`. I think it is at risk of being inadvertently
deleted here and should be moved into `INVOKEAI_ROOT/config` or
`INVOKEAI_ROOT` itself.

**Comment by GreggHelt2** [comment]
> Here is the zipped installer package for anyone who would like to test the wheel-based installation process.
> 
> [InvokeAI-installer-v3.0.0+a1.zip](https://github.com/invoke-ai/InvokeAI/files/11537436/InvokeAI-installer-v3.0.0%2Ba1.zip)

I tested the wheel-based automatic install. Worked great!

Sidenote: I also tried a `pip install -e. `manual install, starting with clone of InovokeAI repo main (as of yesterday). For some reason I got an error building/installing the "lit" package which resulted in crashing out with error. Manually doing `pip install lit` and another round of `pip install -e`. But maybe just a quirk of the evolving head of main?

**Review by GreggHelt2 — APPROVED** [review]

**Comment by psychedelicious** [comment]
@lstein I agree, `invokeai.db` should not be in `outputs`. The runtime root seems like a reasonable place for it, or perhaps in a subfolder called `database`, to shield it from users possibly fiddling with it.

**Review by GreggHelt2 — APPROVED** [review]

**Review by ebr — APPROVED** [review]

**Review by damian0815 — APPROVED** [review]
lgtm

**Review by blessedcoolant — APPROVED** [review]

**Comment by psychedelicious** [comment]
- The application does not create the directories it needs on startup (eg the new `databases` dir). I'm running `invokeai-configure` now, I assume this will create the folder structure.
- The buttons at the bottom of the TUI don't show up until the cursor reaches them if your terminal window is too small.
  - On `Terminal.app` (the default macOS terminal), it resizes the terminal and this problem does not occur.

**Comment by psychedelicious** [comment]
(sorry, I inadvertently hit some github keyboard shortcut that closed the pull request)

**Comment by lstein** [comment]
> * The application does not create the directories it needs on startup (eg the new `databases` dir). ~I'm running `invokeai-configure` now, I assume this will create the folder structure.~ Running `invokeai-configure` did not create the `databases` dir.

On a virgin system it will create the `databases` directory as well as the others. `invokeai-configure` looks for an existing `invokeai.yaml` file and creates the directories if it is missing. I'll have it do a more thorough integrity check of the folder.

 > The buttons at the bottom of the TUI don't show up until the cursor reaches them if your terminal window is too small.
On iTerm2.app (the most popular macOS terminal), the script does not resize itself
On Terminal.app (the default macOS terminal), it resizes the terminal and this problem does not occur
Because my terminal was not large enough, the message saying it was done didn't render, so I was waiting around thinking something was just running slowly.

That's annoying. What I'll do is to launch a new Terminal.app before loading the UI, so that the code can resize the terminal appropriately. What's the command to give at the console to launch Terminal? Is it `Terminal`?

> After starting the downloads, I can still navigate around the TUI and change options. If this doesn't actually change what is being downloaded, the TUI should not allow this

Any changes you make will be reset after the downloads finish. I do hide the INSTALL button while the download is going on in order to discourage people from launching two install processes simultaneously. Completely disabling navigation during downloading may not be possible due to the way that `npyscreen` works.

> The log messages don't show the currently running task. It would be nice to see eg Task (2/15): Install 'controlnet-1.5-canny'

Good suggestion! I also need to add scrollbars to the log field.

**Pull request merged**